File: this-is-coding-test/7.이진_탐색/3.떡볶이_떡_만들기.py
purpose = 6
# tteok = [19, 15, 10, 17]
tteok = [19, 15, 10, 18]


def solution_stack(target, list, start=0, end=None):
    end = max(list) if end is None else end
    cut_line = int((start+end)/2)
    result = 0

    for item in list:
        length = item - cut_line
        result += length if length >= 0 else 0

    print('start:', start, 'end:', end, 'cut_line:', cut_line, 'target:', target, 'result:', result)

    return cut_line if target == result else (
        solution_stack(target, list, cut_line+1, end) if result > target
        else solution_stack(target, list, start, cut_line-1)
    )


# answer_stack = solution_stack(purpose, tteok)
# print(answer_stack)


def solution_queue(target, list):
    start, end = 0, max(list)
    result = None
    test_count, test_limit = 0, 10

    while start < end:
        print(test_count)
        test_count += 1

        cut_line = int((start+end)/2)
        total_quantity = 0


        for item in list:
            length = item - cut_line
            total_quantity += length if length >= 0 else 0

        if total_quantity > target:
            start = cut_line + 1
            result = cut_line
        elif total_quantity < target:
            end = cut_line - 1
        else:
            return cut_line
            break

    return result


answer_queue = solution_queue(purpose, tteok)
print(answer_queue)

File: this-is-coding-test/7.이진_탐색/3.떡볶이_떡_만들기.js
const purpose = 6;
// const items = [19, 15, 10, 17];
const items = [19, 15, 10, 18];

const solveQueue = (target, list) => {
  let start = 0;
  let end = Math.max(...list);
  let result = 0;
  let previous = 0;

  while (start < end) {
    let mid = Math.floor((start + end) / 2);
    console.log(mid);

    const value = items
      .map((item) => (item - mid < 0 ? 0 : item - mid))
      .reduce((cur, acc) => cur + acc);

    if (value > target) {
      if (previous === mid) {
        return mid;
      }
      start = mid;
      previous = result;
      result = mid;
    } else if (value < target) {
      end = mid;
    } else {
      return result;
    }
  }
};

solveQueue(purpose, items);
